ballerina-to-openapi/src/main/java/io/ballerina/openapi/service/mapper/interceptor/Interceptor.java

```diff
@@ -0,0 +1,200 @@
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package io.ballerina.openapi.service.mapper.interceptor;
+
+import io.ballerina.compiler.api.SemanticModel;
+import io.ballerina.compiler.api.symbols.ClassSymbol;
+import io.ballerina.compiler.api.symbols.ResourceMethodSymbol;
+import io.ballerina.compiler.api.symbols.Symbol;
+import io.ballerina.compiler.api.symbols.TypeDefinitionSymbol;
+import io.ballerina.compiler.api.symbols.TypeReferenceTypeSymbol;
+import io.ballerina.compiler.api.symbols.TypeSymbol;
+import io.ballerina.compiler.api.symbols.UnionTypeSymbol;
+import io.ballerina.compiler.syntax.tree.ClassDefinitionNode;
+import io.ballerina.openapi.service.mapper.model.ModuleMemberVisitor;
+import io.ballerina.openapi.service.mapper.utils.MediaTypeUtils;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static io.ballerina.openapi.service.mapper.Constants.BALLERINA;
+import static io.ballerina.openapi.service.mapper.Constants.EMPTY;
+import static io.ballerina.openapi.service.mapper.Constants.HTTP;
+import static io.ballerina.openapi.service.mapper.Constants.NEXT_SERVICE;
+
+/**
+ * This {@link Interceptor} class represents the abstract interceptor service.
+ *
+ * @since 1.9.0
+ */
+public abstract class Interceptor extends Service {
+
+    public enum InterceptorType {
+        REQUEST, REQUEST_ERROR, RESPONSE, RESPONSE_ERROR
+    }
+
+    protected final ClassSymbol serviceClass;
+    private final ClassDefinitionNode serviceClassNode;
+    protected ResourceMethodSymbol resourceMethod = null;
+    protected boolean continueExecution = false;
+    protected boolean hasNilReturn = false;
+    private Interceptor nextInReqPath = null;
+    protected Interceptor nextInResPath = null;
+
+    protected Interceptor(TypeReferenceTypeSymbol typeSymbol, SemanticModel semanticModel,
+                          ModuleMemberVisitor moduleMemberVisitor) throws InterceptorMapperException {
+        super(semanticModel);
+        String name = typeSymbol.getName().orElse("");
+        if (typeSymbol.getName().isPresent() &&
+                moduleMemberVisitor.getInterceptorServiceClassNode(name).isPresent()) {
+            this.serviceClassNode = moduleMemberVisitor.getInterceptorServiceClassNode(name).get();
+        } else {
+            throw new InterceptorMapperException("no class definition found for the interceptor: " + name +
+                    " within the package. Make sure that the interceptor return type is defined with the specific" +
+                    " interceptor class type rather than the generic `http:Interceptor` type and the specific " +
+                    "interceptor class is defined within the package");
```

Comment:
Shall we move these error messages under the diagnostic Message?

Reply:
This error message is specific to the exception I am throwing. We add this as the cause in the diagnostic. This is to imply that the diagnostic is same but the cause for that is different. I prefer not to have different diagnostics when the causes are different